NullPointerException in Component.getLocale() if no I18NProvider configured

### Description of the bug

My application barfed this exception:
```
java.util.concurrent.ExecutionException: java.lang.NullPointerException
        at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)
        at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191)
        at com.vaadin.flow.server.FutureAccess.get(FutureAccess.java:62)
        at com.vaadin.flow.server.VaadinService.runPendingAccessTasks(VaadinService.java:2062)
        at com.vaadin.flow.server.VaadinSession.unlock(VaadinSession.java:712)
        at com.vaadin.flow.server.VaadinService.ensureAccessQueuePurged(VaadinService.java:2025)
        at com.vaadin.flow.server.VaadinService.accessSession(VaadinService.java:1992)
        at com.vaadin.flow.server.VaadinSession.access(VaadinSession.java:1016)
        ...<snip>
Caused by: java.lang.NullPointerException
        at com.vaadin.flow.component.Component.getLocale(Component.java:750)
        at com.vaadin.flow.component.combobox.ComboBoxDataController.lambda$setDataProvider$9fbb9d72$1(ComboBoxDataController.java:457)
        at com.vaadin.flow.component.combobox.ComboBoxDataController.lambda$setDataProvider$13e91f97$1(ComboBoxDataController.java:484)
        at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:176)
        at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1655)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
        at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
        at java.base/java.util.stream.ReduceOps$5.evaluateSequential(ReduceOps.java:257)
        at java.base/java.util.stream.ReduceOps$5.evaluateSequential(ReduceOps.java:248)
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.base/java.util.stream.ReferencePipeline.count(ReferencePipeline.java:605)
        at com.vaadin.flow.data.provider.ListDataProvider.size(ListDataProvider.java:87)
        at com.vaadin.flow.data.provider.DataCommunicator.getDataProviderSize(DataCommunicator.java:923)
        at com.vaadin.flow.data.provider.DataCommunicator.getItemCount(DataCommunicator.java:527)
        at com.vaadin.flow.component.combobox.ComboBoxDataController.refreshAllData(ComboBoxDataController.java:566)
        at com.vaadin.flow.component.combobox.ComboBoxDataController.lambda$setupDataProviderListener$94e23d3c$1(ComboBoxDataController.java:635)
        at com.vaadin.flow.data.provider.AbstractDataProvider$1.accept(AbstractDataProvider.java:64)
        at com.vaadin.flow.data.provider.AbstractDataProvider$1.accept(AbstractDataProvider.java:61)
        at com.vaadin.flow.data.provider.AbstractDataProvider.fireEventForListener(AbstractDataProvider.java:147)
        at com.vaadin.flow.data.provider.AbstractDataProvider.lambda$fireEvent$2(AbstractDataProvider.java:133)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
        at com.vaadin.flow.data.provider.AbstractDataProvider.lambda$fireEvent$3(AbstractDataProvider.java:132)
        at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
        at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
        at java.base/java.util.HashMap$EntrySpliterator.forEachRemaining(HashMap.java:1764)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
        at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
        at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
        at com.vaadin.flow.data.provider.AbstractDataProvider.fireEvent(AbstractDataProvider.java:132)
        at com.vaadin.flow.data.provider.AbstractDataProvider.refreshAll(AbstractDataProvider.java:72)
```

The exception occurred here in `Component.java`:
```java
    protected Locale getLocale() {
        UI currentUi = UI.getCurrent();
        Locale locale = currentUi == null ? null : currentUi.getLocale();
        if (locale == null) {
            List<Locale> locales = getI18NProvider().getProvidedLocales();  // 👈 👈 👈 👈 
            if (locales != null && !locales.isEmpty()) {
                locale = locales.get(0);
            } else {
                locale = Locale.getDefault();
            }
        }
        return locale;
    }   
```
Looks like this method incorrectly assumes that there is always either a current `UI` or a I18NProvider. In the case there is neither, you get an NPE.

Another bug like [#4204](https://github.com/vaadin/flow-components/issues/4204) where code assumes that there is always a current `UI`. This is not correct.

**Yes, there should always be a current locked `VaadinSession`, but when updates occur via background tasks there is no current `UI`.** 

### Expected behavior

There should be no NullPointerExceptions.

### Minimal reproducible example

Invoke `Component.getLocale()` when there is no current `UI` object and no I18NProvider configured.

### Versions

- Vaadin / Flow version: 23.2.7
- Java version: 11
- OS version: Linux

--- a/flow-server/src/main/java/com/vaadin/flow/component/Component.java
+++ b/flow-server/src/main/java/com/vaadin/flow/component/Component.java
@@ -747,7 +747,13 @@ public abstract class Component
         UI currentUi = UI.getCurrent();
         Locale locale = currentUi == null ? null : currentUi.getLocale();
         if (locale == null) {
-            List<Locale> locales = getI18NProvider().getProvidedLocales();
+            final I18NProvider i18NProvider = getI18NProvider();
+            // If a i18nProvider is not defined we should just return the
+            // default locale.
+            if (i18NProvider == null) {
+                return Locale.getDefault();
+            }
+            List<Locale> locales = i18NProvider.getProvidedLocales();
             if (locales != null && !locales.isEmpty()) {
                 locale = locales.get(0);
             } else {

--- a/flow-server/src/test/java/com/vaadin/flow/component/ComponentTest.java
+++ b/flow-server/src/test/java/com/vaadin/flow/component/ComponentTest.java
@@ -306,6 +306,15 @@ public class ComponentTest {
     }
 
     @Test
+    public void getComponentLocale_noCurrentUI_returnsDefaultLocale() {
+        UI.setCurrent(null);
+        Component test = new TestButton();
+        final Locale locale = test.getLocale();
+        Assert.assertEquals("System default locale should be returned",
+                Locale.getDefault(), locale);
+    }
+
+    @Test
     public void getElement() {
         Assert.assertEquals(Tag.DIV,
                 divWithTextComponent.getElement().getTag());

